File: src/main/java/bbarena/model/RollResult.java
package bbarena.model;

import bbarena.model.event.RollModifier;
import bbarena.model.team.AttributeModifier;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class RollResult {

    private final Roll _roll;
    private final int[] _results;
    private final String _why;
    private final String _who;
    private final List<RollModifier> _modifiers;
    private AttributeModifier _attribute = null;
    private Integer _target = null;

    protected RollResult(Roll roll, int[] results, String why, String who) {
        _roll = roll;
        _results = results;
        _why = why;
        _who = who;
        _modifiers = new ArrayList<>();
    }

    public List<RollModifier> getModifiers() {
        return _modifiers;
    }

    public RollResult addModifier(int modifier, String description) {
        addModifier(new RollModifier(modifier, description));
        return this;
    }

    public RollResult addModifier(RollModifier modifier) {
        _modifiers.add(modifier);
        return this;
    }

    public RollResult setAttribute(AttributeModifier attribute) {
        _attribute = attribute;
        return this;
    }

    public RollResult setTarget(Integer target) {
        _target = target;
        return this;
    }

    public Roll getRoll() {
        return _roll;
    }

    public int[] getResults() {
        return _results;
    }

    public Integer getSum() {
        Integer res = Arrays.stream(_results).sum();
        res += _modifiers.stream().mapToInt(value -> value.getModifier()).sum();
        if (_attribute != null) {
            res += _attribute.getMod();
        }
        if (_target != null) {
            res -= _target;
        }
        return res;
    }

    public String getWhy() {
        return _why;
    }

    public String getWho() {
        return _who;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder("Roll [").append(_why).append("] ")
                .append(_roll.getDice()).append(_roll.getType())
                .append(" ")
                .append(Arrays.toString(_results));
        if (_modifiers.size() > 0) {
            builder.append(" + ")
                    .append(_modifiers.toString());
        }
        if (_attribute != null) {
            builder.append(" + ").append(_attribute.getMod()).append(" ").append(_attribute.getType());
        }
        if (_target != null) {
            builder.append(" - ").append(_target).append(" Target");
        }
        builder.append(" = ")
                .append(getSum());
        return builder.toString();
    }
}

File: src/main/kotlin/bbarena/model/RollResult.kt
package bbarena.model

import bbarena.model.event.RollModifier
import bbarena.model.team.AttributeModifier

data class RollResult(val roll: Roll, val results: List<Int>, val why: String, val who: String, val modifiers: MutableList<RollModifier> = mutableListOf(), val attribute: AttributeModifier? = null, val target: Int = 0) {

    val sum: Int
        get() {
            var res: Int = results.sum()
            res += this.modifiers.stream().mapToInt { value -> value.modifier }.sum()
            res += attribute?.mod ?: 0
            res -= target
            return res
        }

    fun addModifier(modifier: Int, description: String): RollResult {
        addModifier(RollModifier(modifier, description))
        return this
    }

    fun addModifier(modifier: RollModifier): RollResult {
        this.modifiers.add(modifier)
        return this
    }

    override fun toString(): String {
        val builder = StringBuilder("Roll [").append(why).append("] ")
                .append(roll.dice).append(roll.type)
                .append(" ")
                .append(results.joinToString())
        if (this.modifiers.size > 0) {
            builder.append(" + ")
                    .append(this.modifiers.toString())
        }
        if (attribute != null) {
            builder.append(" + ").append(attribute.mod).append(" ").append(attribute.attribute)
        }
        builder.append(" - ").append(target).append(" Target")
        builder.append(" = ")
                .append(sum)
        return builder.toString()
    }
}
